src/DustExtinction.jl

```diff
@@ -1,53 +1,132 @@
+
+Redden the given `flux` using the given extinction law at the given wavelength.
+
+If `wave` is `<:Real` then it is expected to be in angstrom and if it is `<:Unitful.Quantity` it will be automatically converted. `Av` is the total extinction value. The extinction law can be a constructed object or just a type. If it is just a type, `kwargs` will be passed to the constructor.
+
+# Examples
+
+```jldoctest
+julia> wave = 3000; flux = 1000;
+
+julia> redden(CCM89, wave, flux; Rv=3.1)
+187.38607779757183
+
+julia> redden(CCM89(Rv=3.1), wave, flux; Av=2)
+35.11354215235764
+```
+
+# See Also
+[`deredden`](@ref)
+"""
+redden(L::Type{<:ExtinctionLaw}, wave, flux; Av = 1, kwargs...) = redden(L(values(kwargs)...), wave, flux; Av = Av)
+redden(law::ExtinctionLaw, wave::Real, flux; Av = 1) = flux * 10^(-0.4 * Av * law(wave))
+redden(law::ExtinctionLaw, wave::Quantity, flux::Real; Av = 1) = redden(law, ustrip(u"Å", wave), flux; Av = Av)
+redden(law::ExtinctionLaw, wave::Quantity, flux::Quantity; Av = 1) = flux * (Av * law(wave))
+
+"""
+    deredden(::ExtinctionLaw, wave, flux; Av=1)
+    deredden(::Type{ExtinctionLaw}, wave, flux; Av=1, kwargs...)
+
+Deredden the given `flux` using the given extinction law at the given wavelength.
+
+If `wave` is `<:Real` then it is expected to be in angstrom and if it is `<:Unitful.Quantity` it will be automatically converted. `Av` is the total extinction value. The extinction law can be a constructed object or just a type. If it is just a type, `kwargs` will be passed to the constructor.
+
+# Examples
+
+```jldoctest
+julia> wave = 3000; flux = 187.386;
+
+julia> deredden(CCM89, wave, flux; Rv=3.1)
+999.9995848273642
+
+julia> deredden(CCM89(Rv=3.1), wave, flux; Av=2)
+5336.573541539394
+```
+
+# See Also
+[`redden`](@ref)
+"""
+deredden(L::Type{<:ExtinctionLaw}, wave, flux; Av = 1, kwargs...) = deredden(L(values(kwargs)...), wave, flux; Av = Av)
+deredden(law::ExtinctionLaw, wave::Real, flux; Av = 1) = flux / 10^(-0.4 * Av * law(wave))
+deredden(law::ExtinctionLaw, wave::Quantity, flux::Real; Av = 1) = deredden(law, ustrip(u"Å", wave), flux; Av = Av)
+deredden(law::ExtinctionLaw, wave::Quantity, flux::Quantity; Av = 1) = flux / (Av * law(wave))
+
+# --------------------------------------------------------------------------------
+# bring in the support
+
+include("deprecate.jl")
+include("color_laws.jl")
+include("dust_maps.jl")
+
+# --------------------------------------------------------------------------------
+# Here be codegen!
+
+# generate unitful support for the following laws
+# this can be removed when julia support is pinned to 1.3 or higher,
+# at which point adding `(l::ExtinctionLaw)(wave)` is possible, until then
+# using this code-gen does the trick but requires manually editing
+# instead of providing support for all <: ExtinctionLaw
+for law in [:CCM89, :OD94, :CAL00, :GCC09, :VCG04]
+    @eval (l::$law)(wavelength::Quantity) = l(ustrip(u"Å", wavelength)) * u"mag"
```

Comment:
Why not simply
```suggestion
for law in [CCM89, OD94, CAL00, GCC09, VCG04]
    (l::law)(wavelength::Quantity) = l(ustrip(u"Å", wavelength)) * u"mag"
```
?  The `@eval` macro is not really needed

Reply:
Because meta programming always starts out as “okay I got this” and always ends with “don’t use @eval” and I never learn my lesson 😂